File: docs/api-audit/probes/cpp/Tensor_04_shape.cpp
// Raw-C++ probe for the Tensor cat-04 (shape / layout) binding-fidelity
// findings. Links against a locally-built libcytnx so it can call the C++
// methods DIRECTLY, bypassing the pybind layer — proving where the Python
// return/identity behavior originates in the C++ signature vs. the binding:
//   * permute_ / reshape_ return Tensor& (&*this) in C++, so the Python self-
//     return the pybind lambdas give (`return &self.permute_(...)`) is FAITHFUL
//     (finding T-S2);
//   * permute (no underscore) returns a DISTINCT object that SHARES data — the
//     shared-data view the Python probe observes (finding T-S1);
//   * contiguous_ returns `Tensor` BY VALUE (a *this copy sharing storage), NOT
//     Tensor& — so Python's distinct-object return is FAITHFUL to C++; the
//     broken self-return convention is ROOTED IN THE C++ SIGNATURE, not the
//     binding (finding T-S5);
//   * flatten_ returns `void` in C++ — so Python's None is FAITHFUL; the missing
//     self-return is ROOTED IN THE C++ SIGNATURE (finding T-S7).
#include "cytnx.hpp"
#include <iostream>
#include <type_traits>
#include <vector>
using namespace cytnx;

static int fails = 0;
static void report(const std::string& claim, bool ok) {
  std::cout << (ok ? "[PASS] " : "[FAIL] ") << claim << std::endl;
  if (!ok) ++fails;
}

// Static (compile-time) facts about the C++ return TYPES — the root cause of
// the Python return-convention differences. These are the load-bearing
// binding-fidelity facts a runtime `is`-check alone cannot pin down.
static_assert(
    std::is_same<decltype(std::declval<Tensor&>().permute_(
                     std::declval<const std::vector<cytnx_uint64>&>())),
                 Tensor&>::value,
    "C++ permute_ must return Tensor& (self)");
static_assert(
    std::is_same<decltype(std::declval<Tensor&>().reshape_(
                     std::declval<const std::vector<cytnx_int64>&>())),
                 Tensor&>::value,
    "C++ reshape_ must return Tensor& (self)");
static_assert(
    std::is_same<decltype(std::declval<Tensor&>().contiguous_()), Tensor>::value,
    "C++ contiguous_ returns Tensor BY VALUE (not Tensor&) — root of T-S5");
static_assert(
    std::is_same<decltype(std::declval<Tensor&>().flatten_()), void>::value,
    "C++ flatten_ returns void — root of T-S7");

int main() {
  report("T-S2: C++ permute_ has return type Tensor& (static_assert) — the "
         "pybind lambda's `return &self.permute_(...)` self-return is FAITHFUL",
         true);
  report("T-S2: C++ reshape_ has return type Tensor& (static_assert) — its "
         "pybind self-return is FAITHFUL",
         true);
  report("T-S5: C++ contiguous_ has return type Tensor BY VALUE, NOT Tensor& "
         "(static_assert) — so Python's distinct-object return is FAITHFUL to "
         "C++; the broken in-place self-return is ROOTED IN THE C++ SIGNATURE",
         true);
  report("T-S7: C++ flatten_ has return type void (static_assert) — so Python's "
         "None return is FAITHFUL to C++; the missing self-return is ROOTED IN "
         "THE C++ SIGNATURE",
         true);

  // T-S2 (runtime): C++ permute_ / reshape_ return &*this — the actual self.
  {
    Tensor t = arange(24).reshape(2, 3, 4);
    Tensor& r = t.permute_(std::vector<cytnx_uint64>{2, 0, 1});
    report("T-S2: C++ permute_ permutes in place and returns &*this (&r == &t); "
           "shape becomes [4,2,3]",
           &r == &t && t.shape() == std::vector<cytnx_uint64>{4, 2, 3});
  }
  {
    Tensor t = arange(24).reshape(2, 3, 4);
    Tensor& r = t.reshape_(std::vector<cytnx_int64>{6, 4});
    report("T-S2: C++ reshape_ reshapes in place and returns &*this (&r == &t); "
           "shape becomes [6,4]",
           &r == &t && t.shape() == std::vector<cytnx_uint64>{6, 4});
  }

  // T-S1: C++ permute (no underscore) returns a DISTINCT object whose data is
  // SHARED with the source — the shared-data view the Python probe observes.
  {
    Tensor t = arange(24).reshape(2, 3, 4);
    Tensor v = t.permute(std::vector<cytnx_uint64>{2, 0, 1});
    t.at<double>({0, 0, 0}) = 9.0;  // mutate source data
    bool shared = double(v.at<double>({0, 0, 0})) == 9.0;
    report("T-S1: C++ permute (pure) returns a distinct object with reordered "
           "legs but SHARED data (mutating the source shows through the permute)",
           (&v != &t) && v.shape() == std::vector<cytnx_uint64>{4, 2, 3} && shared);
  }

  // T-S5 (runtime): the by-value contiguous_ result shares storage with the
  // receiver (it is *this copied — a shared-data wrapper) and coalesces the
  // receiver in place.
  {
    Tensor t = arange(24).reshape(2, 3, 4).permute(std::vector<cytnx_uint64>{2, 0, 1});
    bool was_noncontig = !t.is_contiguous();
    Tensor r = t.contiguous_();  // by-value copy of *this
    report("T-S5: C++ contiguous_ coalesces the receiver in place (now "
           "contiguous) and its BY-VALUE result shares storage with the "
           "receiver (same_data) — a shared-data wrapper, not &*this",
           was_noncontig && t.is_contiguous() && r.same_data(t));
  }

  // T-S1/T-S8: C++ contiguous (no underscore) returns a distinct object; on a
  // non-contiguous tensor it is an INDEPENDENT copy (does not share data).
  {
    Tensor t = arange(24).reshape(2, 3, 4).permute(std::vector<cytnx_uint64>{2, 0, 1});
    Tensor c = t.contiguous();
    report("T-S8: C++ contiguous (no underscore, the raw method bound as "
           "make_contiguous) returns a distinct, contiguous object; on a "
           "non-contiguous tensor it does NOT share data with the source",
           (&c != &t) && c.is_contiguous() && !c.same_data(t));
  }

  // T-S7 (runtime): C++ flatten_ collapses the receiver in place; being void it
  // cannot return self at all.
  {
    Tensor t = arange(24).reshape(2, 3, 4);
    t.flatten_();
    report("T-S7: C++ flatten_ collapses the receiver to rank 1 in place "
           "(shape [24]); being void it returns nothing",
           t.shape() == std::vector<cytnx_uint64>{24});
  }

  // T-S6: C++ flatten (no underscore) returns an INDEPENDENT 1-D copy
  // (clone + contiguous_ + reshape_), leaving the source unchanged.
  {
    Tensor t = arange(24).reshape(2, 3, 4);
    Tensor f = t.flatten();
    report("T-S6: C++ flatten (pure) returns an independent rank-1 copy "
           "(shape [24], does not share data) while the source stays rank 3",
           f.shape() == std::vector<cytnx_uint64>{24} && !f.same_data(t) &&
               t.shape() == std::vector<cytnx_uint64>{2, 3, 4});
  }

  std::cout << (fails ? "cpp probe FAILED" : "cpp probe ok") << std::endl;
  return fails ? 1 : 0;
}

File: docs/api-audit/probes/Tensor_04_shape.py
"""Behavioral probe for the Tensor audit, category 04 (shape / layout),
verified against the installed cytnx==1.1.0 wheel (NOT source-inferred).

Every runtime claim in docs/api-audit/Tensor/04-shape-layout.md is backed by a
report(...) assertion here. Members covered: permute, permute_, reshape,
reshape_, contiguous, contiguous_, flatten, flatten_, and the leaked raw
make_contiguous shim.

The binding-fidelity findings (T-S3 *args erasure, T-S5 contiguous_ identity
drop, T-S7 flatten_ None) additionally have their raw-C++ side verified by
probes/cpp/Tensor_04_shape.cpp against a source-built libcytnx.

Run: source tools/env.sh && $PY docs/api-audit/probes/Tensor_04_shape.py
"""
import sys, os, inspect

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "tools"))
import cytnx
from probe_helper import report, returns_view

Tensor = cytnx.Tensor

report("wheel under test is 1.1.0 (runtime, not source-inferred)",
       cytnx.__version__ == "1.1.0")


def _mk():
    """A fresh, contiguous rank-3 Double tensor [2,3,4] = arange(24)."""
    return cytnx.arange(24).reshape(2, 3, 4)


# =========================================================================
# T-S1: permute / reshape return a DISTINCT object that SHARES storage with
# the receiver — a metadata-only VIEW, not a copy (numpy divergence for those
# expecting reshape to sometimes copy, but here it aliases).
# =========================================================================
src = _mk()
p = src.permute(2, 0, 1)
report("permute returns a DISTINCT object (pure — the receiver is untouched)",
       p is not src)
report("permute returns a shared-data VIEW (same_data(src) is True)",
       p.same_data(src) is True)
report("permute VIEW is live: mutating the source shows through the permute",
       returns_view(_mk, lambda s: s.permute(2, 0, 1),
                    lambda h: h.__setitem__((0, 0, 0), 777),  # write through view
                    lambda s: float(s[0, 0, 0].item())) is True)

src = _mk()
r = src.reshape(6, 4)
report("reshape returns a DISTINCT object (pure)", r is not src)
report("reshape returns a shared-data VIEW (same_data(src) is True)",
       r.same_data(src) is True)
report("reshape VIEW is live: a write through the reshape is visible on the source",
       returns_view(_mk, lambda s: s.reshape(6, 4),
                    lambda h: h.__setitem__((0, 0), 555),
                    lambda s: float(s[0, 0, 0].item())) is True)

# =========================================================================
# T-S2: permute_ / reshape_ are correct in-place methods that return SELF
# (the same Python object) — chainable. C++ returns Tensor& ; the pybind
# lambdas return &self (tensor_py.cpp:179-198).
# =========================================================================
a = _mk()
ret = a.permute_(2, 0, 1)
report("permute_ permutes in place and returns SELF (ret is a)", ret is a)
report("permute_ actually reordered the receiver's shape to [4,2,3]",
       a.shape() == [4, 2, 3])

b = _mk()
ret = b.reshape_(6, 4)
report("reshape_ reshapes in place and returns SELF (ret is b)", ret is b)
report("reshape_ actually reshaped the receiver to [6,4]", b.shape() == [6, 4])

# =========================================================================
# T-S3: permute / reshape (and their _ forms) are bound as (*args) pybind
# lambdas — the typed C++ (mapper/new_shape, ...) signature is ERASED, so
# inspect.signature() raises and the docstring reads "(*args)".
# =========================================================================
for name in ("permute", "permute_", "reshape", "reshape_"):
    m = getattr(Tensor, name)
    raised = False
    try:
        inspect.signature(m)
    except ValueError:
        raised = True
    report(f"{name} is bound as a (*args) pybind lambda — inspect.signature() raises ValueError",
           raised)
    report(f"{name}'s docstring exposes the erased (*args) signature, not a typed one",
           "*args" in (m.__doc__ or "").splitlines()[0])

# =========================================================================
# T-S4: contiguous is a conti.py wrapper (Tensor_conti.py:50-55) that
# SHORT-CIRCUITS to self when already contiguous, else forwards to the leaked
# raw make_contiguous shim (a distinct, contiguous COPY).
# =========================================================================
c = _mk()
report("a freshly-built tensor is contiguous", c.is_contiguous() is True)
report("contiguous() short-circuits to SELF when the tensor is already contiguous",
       c.contiguous() is c)

nc = _mk().permute(2, 0, 1)
report("a permuted tensor is NON-contiguous", nc.is_contiguous() is False)
cc = nc.contiguous()
report("contiguous() on a non-contiguous tensor returns a DISTINCT object",
       cc is not nc)
report("contiguous() on a non-contiguous tensor returns an INDEPENDENT copy "
       "(same_data is False) that is contiguous",
       cc.same_data(nc) is False and cc.is_contiguous() is True)

# =========================================================================
# T-S5: contiguous_ is the in-place layout coalesce, BUT the binding returns a
# DISTINCT shared-data wrapper, NOT self — unlike permute_/reshape_. Root
# cause is the C++ signature `Tensor contiguous_()` (returns *this BY VALUE,
# not Tensor&), bound as a plain method pointer (tensor_py.cpp:193). The raw
# C++ side is verified by the C++ probe.
# =========================================================================
d = _mk().permute(2, 0, 1)
report("contiguous_ target is non-contiguous before the call", d.is_contiguous() is False)
ret = d.contiguous_()
report("contiguous_ coalesces the RECEIVER's storage in place (d is now contiguous)",
       d.is_contiguous() is True)
report("contiguous_ returns a DISTINCT object, NOT self (ret is not d) — the "
       "in-place self-return convention is broken (contrast permute_/reshape_)",
       ret is not d)
report("contiguous_'s returned wrapper still SHARES data with the receiver "
       "(same_data(d) is True) — a shared-data view, not an independent copy",
       ret.same_data(d) is True)

# =========================================================================
# T-S6: flatten is a PURE 1-D collapse returning an INDEPENDENT COPY
# (C++ flatten() = clone + contiguous_ + reshape_, Tensor.hpp:1416).
# =========================================================================
e = _mk()
fl = e.flatten()
report("flatten returns a DISTINCT object collapsed to rank 1 ([24])",
       fl is not e and fl.shape() == [24])
report("flatten returns an INDEPENDENT COPY (same_data(src) is False) — the "
       "receiver is untouched (still rank 3)",
       fl.same_data(e) is False and e.shape() == [2, 3, 4])

# =========================================================================
# T-S7: flatten_ is the in-place 1-D collapse — but it returns None, NOT self
# (v1 Tensor C4). Root cause is the C++ signature `void flatten_()`
# (Tensor.hpp:1430), bound as a plain method pointer (tensor_py.cpp:191).
# =========================================================================
g = _mk()
ret = g.flatten_()
report("flatten_ collapses the RECEIVER to rank 1 in place ([24])", g.shape() == [24])
report("flatten_ returns None, NOT self — breaks the in-place self-return "
       "convention its siblings permute_/reshape_ keep (v1 C4)",
       ret is None)

# =========================================================================
# T-S8: the raw make_contiguous shim LEAKS into public dir(Tensor). It is the
# raw C++ contiguous() bound under a renamed name (tensor_py.cpp:192, comment
# "this will be rename by python side conti"); unlike the public contiguous()
# wrapper it does NOT short-circuit — on an already-contiguous tensor it
# returns a NEW wrapper that still SHARES storage (a view).
# =========================================================================
report("the raw make_contiguous shim LEAKS into public dir(Tensor)",
       "make_contiguous" in dir(Tensor))
h = _mk()
mk = h.make_contiguous()
report("make_contiguous does NOT short-circuit: on an already-contiguous "
       "tensor it returns a DISTINCT wrapper (mk is not h)",
       mk is not h)
report("make_contiguous's wrapper still SHARES storage (same_data is True) — a "
       "view, unlike contiguous()'s identity short-circuit",
       mk.same_data(h) is True)

print("Tensor 04 probe ok")
